add ical feed [#11]

simon-weber opened the pull request:
I added the ical feed at `/ical`. It'll create all-day events on Google Calendar.

Three things to bring your attention to:
- I added `icalendar` as a dependency, which in turn depends on a couple of other packages. If you'd prefer not to use that, I could rewrite it using a template instead
- see line 346 for how I handled releases without days
- presently, there's not actually a link to the feed. I was thinking of adding a little calendar icon next to the rss feed icon, but I wanted to see what you thought first

simon-weber committed:
add ical feed [#11]

alexkay commented on app/views.py:
```diff
@@ -24,6 +24,7 @@
 from django.http import HttpResponse, HttpResponseForbidden, HttpResponseNotFound
 from django.shortcuts import redirect, render
 from django.views.decorators.cache import cache_control
+from icalendar import Calendar, Event
```
Could you indeed rewrite this without the icalendar dependency. Let me know if it's too involved.

alexkay commented on app/views.py:
```diff
@@ -311,6 +312,53 @@ def feed(request):
+    releases = list(ReleaseGroup.get(user=profile.user, limit=LIMIT, offset=0, feed=True))
+
+    cal = Calendar()
+    cal.add('prodid', '-//Muspy releases//mxm.dk//')
```
Why mxm.dk?

simon-weber replied:
That's. . .a very good question, haha. I copied that from a template thinking it had some kind of significance.

Apparently the usual form for this is `-//company//product//language`. I'll be sure to change this.

alexkay commented on app/views.py:
```diff
@@ -311,6 +312,53 @@ def feed(request):
+        event.add('dtstart', event_date)
+        event.add('dtend', event_date + timedelta(days=1))
+
+        #event['uid'] = '20050115T101010/27346262376@mxm.dk'
```
Remove this comment.

alexkay commented on app/views.py:
```diff
@@ -311,6 +312,53 @@ def feed(request):
+# copy-paste from feed -- you probably want to factor that out?
+def ical(request):
+    user_id = request.GET.get('id', '')
+    if user_id.isdigit():
```
This logic is not needed here. It exists in the /feed handler for legacy reasons (in the early days the feed links were by numeric IDs).

alexkay commented on app/views.py:
```diff
@@ -311,6 +312,53 @@ def feed(request):
+            # arbitrarily set the release towards the end of the month.
+            # hopefully, the date will be clarified before then, but this
+            # will ensure it's not missed on the calendar
+            day = 28
```
You could also set it to the last day of that month. It would be just a couple of lines of datetime trickery.

alexkay commented:
Hey Simon, looks good overall, thank you for working on this! I added a few comments in the code.

> presently, there's not actually a link to the feed. I was thinking of adding a little calendar icon next to the rss feed icon, but I wanted to see what you thought first

Sounds good.

simon-weber commented:
Great; thanks for the comments. I'll fix this up in the next day or two.

simon-weber committed:
address pr comments; remove ical dependency

simon-weber commented:
I addressed all your comments. Here are a few more notes:
- I never got static files to show up locally, so be sure to double check the styling of the calendar icon I added (I'm willing to bet it's too big or something)
- the image is from [here](http://www.iconarchive.com/show/oxygen-icons-by-oxygen-icons.org/Actions-view-calendar-day-icon.html), and apparently safe for commercial use. You can check out some alternatives [here](http://www.iconarchive.com/search?q=calendar&license=commercial&page=1&res=small).
- I use release.id / userid to make the eventids unique inside muspy
- I've tested importing the resulting file on Google Calendar and it works -- I haven't tested add-by-url/syncing

simon-weber commented:
bump =)

alexkay commented:
Sorry for delay Simon! The changes look good. I'll merge, test and deploy later this week.

simon-weber commented:
Awesome; thanks!

alexkay commented:
Tested, merged, pushed and deployed. Thank you Simon and sorry it took so long!

simon-weber commented:
:sparkling_heart:

simon-weber commented:
Oh, sweet -- I even get new event notifications now:

```
Blood Stain Child - LAST STARDUST
When
Wed Jan 29, 2014
Calendar
Muspy Releases
Invitation from Google Calendar

You are receiving this email at the account <snip>@gmail.com
because you are subscribed for new event notifications on calendar Muspy Releases.

To stop receiving these notifications, please log in to https://www.google.com/calendar/
and change your notification settings for this calendar.
```

Thanks again for the merge =)

alexkay commented:
No worries, thank you for implementing it! :)
